File: java-midterm/src/main/java/com/pht/service/ProductService.java
package com.pht.service;

import com.pht.dro.response.ProductResponse;
import com.pht.dto.CustomPaging;
import com.pht.dto.request.FilterRequest;
import com.pht.dto.request.ProductRequest;

public interface ProductService {
	CustomPaging<ProductResponse> getByFilter(FilterRequest filter);
	
	ProductResponse getById(long id);
	
	ProductResponse create(ProductRequest request);
	
	ProductResponse update(ProductRequest request, long id);
	
	void delete(long id);
}

File: midterm/src/app/services/fetch/product.service.ts
import { ProductFilter } from './../../models/product';
import { Injectable } from '@angular/core';
import { BaseService } from './base.service';
import { Observable } from 'rxjs';
import { Paging } from '../../models/paging';
import { ProductResponse } from '../../models/product';

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  constructor(private readonly baseService: BaseService) { }

  getAll(filter: ProductFilter): Observable<Paging<ProductResponse>> {
    return this.baseService.post<Paging<ProductResponse>>('/api/products/filter/', filter);
  }

  getById(id: number): Observable<ProductResponse> {
    return this.baseService.get<ProductResponse>(`/api/products/${id}/`);
  }

  create(formData: FormData): Observable<ProductResponse> {
    return this.baseService.post<ProductResponse>('/api/products/', formData, true);
  }

  update(id: number, formData: FormData): Observable<any> {
    return this.baseService.put<ProductResponse>(`/api/products/${id}/`, formData, true);
  }

  delete(id: number): Observable<any> {
    return this.baseService.delete<any>(`/api/products/${id}/`);
  }
}
